### DISCUSSION TITLE
[Documentation] Stage 2 Uses Full-Parameter Fine-Tuning

### DISCUSSION THREAD

**Discussion by WarriorHanamy** [WarriorHanamy/parkour#2] [General]
# Stage 2 Training Strategy: Full-Parameter Fine-Tuning

## Executive Summary

This document analyzes codebase to determine the fine-tuning strategy used in **Stage 2** (RL Fine-tuning with Hard Dynamics Constraints) of the Robot Parkour Learning framework.

**Finding**: Stage 2 uses **full-parameter fine-tuning** (all model parameters are updated) rather than partial or parameter-efficient fine-tuning (PEFT) techniques.

---

### 1. Loading Mechanism (Stage 1 → Stage 2)

**File**: `legged_gym/legged_gym/utils/task_registry.py` (lines 168-175)

```python
resume = train_cfg.runner.resume
if resume:
    # Load previously trained model
    resume_path = get_load_path(log_root, load_run=train_cfg.runner.load_run, ...)
    print(f"Loading model from: {resume_path}")
    if save_cfg:
        shutil.copyfile(resume_path, os.path.join(log_dir, os.path.basename(resume_path)))
    runner.load(resume_path)
```

**Analysis**: Stage 2 loads the entire pre-trained model from Stage 1 checkpoint without any parameter selection or filtering.

---

### 2. No Parameter Freezing Logic

**File**: `rsl_rl/rsl_rl/runners/on_policy_runner.py` (lines 258-268)

```python
def load(self, path, load_optimizer=True):
    loaded_dict = torch.load(path)
    if self.cfg.get("ckpt_manipulator", False):
        # Optional checkpoint manipulation (not used in Stage 2)
        print("\033[1;36m Warning: using a hacky way to load the model. \033[0m")
        loaded_dict = getattr(ckpt_manipulator, self.cfg["ckpt_manipulator"])(
            loaded_dict,
            self.alg.state_dict(),
        )
        print("\033[1;36m Done: using a hacky way to load the model. \033[0m")
    self.alg.load_state_dict(loaded_dict)  # ← Loads ALL parameters
    self.current_learning_iteration = loaded_dict['iter']
    ...
```

**File**: `rsl_rl/rsl_rl/algorithms/ppo.py` (lines 236-242)

```python
def load_state_dict(self, state_dict):
    self.actor_critic.load_state_dict(state_dict["model_state_dict"])  # ← All parameters
    if "optimizer_state_dict" in state_dict:
        self.optimizer.load_state_dict(state_dict["optimizer_state_dict"])
    ...
```

**Analysis**:
- ❌ No `requires_grad = False` found
- ❌ No parameter masking or selective loading
- ❌ No frozen layer logic in Stage 2 pipeline
- `ckpt_manipulator.py` only contains `replace_encoder0()` for specific use cases (not Stage 2)

---

### 3. Optimizer Updates All Parameters

**File**: `rsl_rl/rsl_rl/algorithms/ppo.py` (line 70)

```python
def __init__(self, ...):
    ...
    self.optimizer = getattr(optim, optimizer_class_name)(
        self.actor_critic.parameters(),  # ← All parameters passed to optimizer
        lr=learning_rate
    )
```

**Analysis**: The optimizer receives `self.actor_critic.parameters()`, which includes **all trainable parameters** (actor network, critic network, RNN hidden states, etc.).

---

### 4. Stage 2 Configuration Characteristics

**File**: `legged_gym/legged_gym/envs/a1/a1_field_config.py` (lines 272-302)

```python
class A1FieldCfgPPO( A1RoughCfgPPO ):
    class algorithm( A1RoughCfgPPO.algorithm ):
        entropy_coef = 0.01  # ← Lower exploration
        clip_min_std = 1e-12

    class policy( A1RoughCfgPPO.policy ):
        rnn_type = 'gru'  # ← Uses RNN

    class runner( A1RoughCfgPPO.runner ):
        policy_class_name = "ActorCriticRecurrent"
        experiment_name = "field_a1"
        resume = True  # ← Load Stage 1 model

        run_name = "".join([
            "WalkForward",
            # ... config string ...
            ("_from" + "_".join(load_run.split("/")[-1].split("_")[:2]) if resume else "_noResume"),
        ])
        max_iterations = 5000  # ← Shorter training
        save_interval = 500
```

**Key Characteristics**:

| Configuration | Value | Purpose |
|---------------|--------|---------|
| `resume` | `True` | Load pre-trained model |
| `entropy_coef` | `0.01` | Reduce exploration (fine-tuning) |
| `max_iterations` | `5000` | Shorter training duration |
| `rnn_type` | `'gru'` | Recurrent network |
| `algorithm` | `PPO` | Standard PPO |

---

## Three-Stage Training Comparison

| Stage | Algorithm | Strategy | Training Time | Key Parameters | Purpose |
|-------|-----------|-----------|---------------|----------------|---------|
| **Stage 1** | PPO | Train from scratch | ~20,000+ iterations | `entropy_coef=0.01` | Learn basic walking skills |
| **Stage 2** | PPO | **Full-parameter fine-tuning** | ~5,000 iterations | `entropy_coef=0.01` | Adapt to specific parkour skills |
| **Stage 3** | TPPO | Teacher-Student distillation | Variable | `distillation_loss_coef=50.0` | Learn vision-based policy |

---

### 1. Task Similarity

- Stage 2 skills (tilt, jump, crawl, leap, down) are **highly related** to walking
- All skills share the same locomotion primitives
- Full fine-tuning preserves learned representations while adapting to new constraints

### 2. Architecture Consistency

- All stages use the same `ActorCriticRecurrent` architecture
- Same observation space structure
- Same action space (12 DOF control)
- Enables seamless parameter transfer

### 3. Training Efficiency

- **Short training time** (5,000 iterations) prevents catastrophic forgetting
- **Lower entropy** reduces exploration, focusing on policy refinement
- Pre-trained weights provide strong initialization, speeding up convergence

### 4. Hardware Compatibility

- Single checkpoint file contains all parameters
- Simpler deployment (no need for multi-stage loading)
- Consistent inference pipeline across all skills

### 5. Domain Randomization

- Stage 2 applies realistic constraints (hard dynamics, physical obstacles)
- Full fine-tuning allows the policy to adapt perception-action loops
- Sensor latency and motor dynamics are learned end-to-end

---

### Example 1: Tilt Skill Configuration

**File**: `legged_gym/legged_gym/envs/a1/a1_tilt_config.py`

```python
class A1TiltCfgPPO( A1FieldCfgPPO ):
    class algorithm( A1FieldCfgPPO.algorithm ):
        entropy_coef = 0.0  # ← Even lower for fine-tuning
        clip_min_std = 0.2

    class runner( A1FieldCfgPPO.runner ):
        policy_class_name = "ActorCriticRecurrent"
        experiment_name = "field_a1"
        resume = True  # ← Load from Stage 1
        load_run = osp.join(logs_root, "field_a1_noTanh_oracle",
                          "Oct11_12-24-22_Skills_tilt_...")

        run_name = "".join([
            "Skills_",
            ("Multi" if len(A1TiltCfg.terrain.BarrierTrack_kwargs["options"]) > 1
             else A1TiltCfg.terrain.BarrierTrack_kwargs["options"][0]),
            ("_noResume" if not resume else
             "from" + "_".join(load_run.split("/")[-1].split("_")[:2])),
        ])
        max_iterations = 20000
        save_interval = 500
```

### Example 2: Jump Skill Configuration

**File**: `legged_gym/legged_gym/envs/a1/a1_jump_config.py`

```python
class A1JumpCfgPPO( A1FieldCfgPPO ):
    class algorithm( A1FieldCfgPPO.algorithm ):
        entropy_coef = 0.0

    class runner( A1FieldCfgPPO.runner ):
        resume = True
        load_run = osp.join(logs_root, "field_a1", "...")
```

---

## Alternative Approaches (Not Used)

| Approach | Description | Why Not Used in Stage 2 |
|----------|-------------|------------------------|
| **LoRA** | Low-rank adaptation | Overkill for simple locomotion tasks |
| **Adapter Layers** | Add small trainable layers | Same architecture constraints |
| **Partial Freezing** | Freeze encoder, train head | No clear separation in shared representation |
| **Gradient Checkpointing** | Reduce memory usage | Not needed for 5,000 iterations |

---

### For Researchers

1. **Understand the baseline**: Stage 2 is full-parameter fine-tuning
2. **Compare with PEFT**: If comparing methods, ensure fair baselines
3. **Training budget**: 5,000 iterations is sufficient for adaptation from good initialization

### For Developers

1. **Add to README**: Document three-stage training methodology
2. **Code comments**: Add notes in config files explaining fine-tuning strategy
3. **Parameter tracking**: Consider logging parameter changes during fine-tuning

### For Reproduction

```bash
# Stage 1: Train walking policy
python -m legged_gym.scripts.train --task a1_field

# Stage 2: Fine-tune for tilt skill
python -m legged_gym.scripts.train --task a1_tilt

# Stage 3: Distill to vision policy
python -m legged_gym.scripts.train --task a1_field_distill
```

---

## Conclusion

**Stage 2 uses full-parameter fine-tuning**, which is an appropriate design choice for the Robot Parkour Learning framework:

✅ Efficient adaptation from pre-trained walking policy
✅ Maintains learned locomotion primitives
✅ Prevents catastrophic forgetting with short training
✅ Consistent architecture across all stages
✅ Simple deployment and inference pipeline

This design is well-suited for learning parkour skills that build upon foundational walking capabilities.

---

## References

- **Original Paper**: [Robot Parkour Learning (CoRL 2023)](https://robot-parkour.github.io/)
- **Repository**: https://github.com/ZiwenZhuang/parkour
- **Key Files Analyzed**:
  - `legged_gym/legged_gym/utils/task_registry.py`
  - `rsl_rl/rsl_rl/algorithms/ppo.py`
  - `rsl_rl/rsl_rl/runners/on_policy_runner.py`
  - `legged_gym/legged_gym/envs/a1/a1_*_config.py`

---

*Document created: February 27, 2026*
*Codebase version: Analysis of current master branch*
